### PR TITLE
[New] Msg plugin / _prepend_tree update.

### PR DESCRIPTION
**Pull request by bucaran**
```
   __ _  ___ ___ _
  /  ' \(_-</ _ `/
 /_/_/_/___/\_, /
           /___/ v0.1.0

NAME
     msg - technicolor message printer

SYNOPSIS
     msg [-sn] @<fg:bg> [style]<text>[style]

OPTIONS
     -s
         Do not separate arguments with spaces.
     -n
         Do not output a newline at the end of the message.

STYLES
     _text_                 Bold
     __text__               Underline
     ___text___             Bold and Underline
     /directory/            Directories
     [url]                  Links
     \n                     Line Break
     \t                     Tab Space

     @<fg:bg> fg=bg=RGB|RRGGBB|name|random
       RGB value can be any three to six hex digit or color name.
       e.g, @0fb, @tomato, @random, @error, @ok.

AUTHORS
     @bucaran (Jorge Bucaran)

v. 0.1.0
```

Some utility functions are declared inside `msg.fish` in order to simplify and separate concerns, like reducing the amount of `sed` calls in the wild. The following provides some documentation for local utilities to `msg`.

### `function --no-scope-shadowing msg._.set.color`
Set style, fg/bg colors and reset the color again. This function modifies variables in the parent scope.
`params [<fg>] [<bg>] [<style>]`

### `function msg._.str.has`
True if substring exists in string. Used to cascade apply `-sn` options when `msg` is called.
`params <substring> <string>`

### `function msg._.str.grab`
Extract string between left and right separators of variable length. Used to grab the part inside the style separators. `_bold_, __underline__, ___bold_and_underline___, [urls], /directories/`
`params <left-sep> [<right-sep>] <string>`

### `function msg._.random.color`
Print a random `RRGGBB` hexadecimal color. 

### `function msg._.get.color`
Translate color names to valid RRGGBB hexadecimal value. This function takes the bulk of the code in the module as it checks for all 256 color names.
`params <color|random>`

See #283 and #285 for more info.

### CODE REVIEW

**Comment by bucaran** [comment]
This PR also shows all my previous commits because the branches were locally merged. This is however a new PR fully in sync with your most recent master.

File: plugins/msg/msg.fish [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+        echo -en (msg._.set.color)$token$reset
+        [ -z $ws -o $argv[-1] = $token ]
+          or echo -n $space
+    end
+  end
+end
+
+# True if substring exists in string.
+# @params <substring> <string>
+function msg._.str.has
```
Comment (bpinto) [inline plugins/msg/msg.fish]: What is the `_` for?

What do you think about `function _msg_string_has`?

File: plugins/msg/msg.fish [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+  printf $argv[2] | grep -q $argv[1]
+end
+
+# Extract string between left and right separators of variable length.
+# @params <left-sep> [<right-sep>] <string>
+function msg._.str.grab
```
Comment (bpinto) [inline plugins/msg/msg.fish]: `_msg_string_extract`

File: plugins/msg/msg.fish [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+# Print a random RRGGBB hexadecimal color.
+function msg._.random.color
+  printf "%02x%02x%02x" (math (random) "%" 255) \
+                        (math (random) "%" 255) \
+                        (math (random) "%" 255)
+end
+
+# Translate color names to valid RRGGBB hexadecimal value.
+# @params <color|random>
+function msg._.get.color
```
Comment (bpinto) [inline plugins/msg/msg.fish]: `_msg_color_parse`

File: plugins/msg/msg.fish [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+   	case whitesmoke
+      printf F5F5F5
+   	case white
+      printf FFFFFF
+    case "*"
+      printf $argv[1]
+  end
+end
+
+function msg._.test
```
Comment (bpinto) [inline plugins/msg/msg.fish]: When is this used?

File: plugins/msg/msg.fish (hunk 1 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: I use `_` to mean local functions / utilities.

I really like the `module_name_._.command` convention to name functions given that `.` are valid function names (but not variables) in Fish. I wouldn't mind a lot changing this, but I am writing another plugin right now that is pretty large using this convention consistently and it's really useful when reusing default keywords like `set` or `contains` as they are automatically highlighted by my editor. So I could for instance `msg._.str.contains` and `contains` would be highlighted adding readability without interfering with anything.

Comment (bpinto) [inline plugins/msg/msg.fish]: :+1: 

I don't have a strong opinion on this. Let's try to embrace this convention and see how it goes. :)

File: plugins/msg/msg.fish (hunk 4 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: I used it just for testing it, it's _not_ strictly necessary.

Comment (bpinto) [inline plugins/msg/msg.fish]: Have you heard about [fish-tank](https://github.com/terlar/fish-tank)? Might be a better solution.

File: plugins/msg/msg.fish (hunk 1 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: Cool. If you don't mind, I will leave the `._.` as well, as it's shorter than, e.g., `msg.utils.command` which lets me stay under 80 characters per column in a few long one-liners. I

Comment (bpinto) [inline plugins/msg/msg.fish]: :+1:

File: plugins/msg/msg.fish (hunk 4 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: Yes, but I am not exactly sure how I could test `msg`. For example, there is no `get_color` in Fish so how can I confidently know what color was set by `set_color`? Also, according to `set_color`'s manual, not all terminal emulators support all the color / style features. I will definitely look into this later though.

File: plugins/msg/msg.fish (hunk 1 above)
Comment (bucaran) [inline plugins/msg/msg.fish]: Thanks.

**Pull request closed without merging**